// serverless-base-util/database/Database.js
const { Sequelize, Op } = require("sequelize");

module.exports = {
    isConnectionCreated: false,
    errorMessage: null,
    db: {},
    init: async function () {
        try {
            //To create a MySQL connection
            const sequelize = new Sequelize(
                process.env.DATABASE_SCHEMA,
                process.env.DATABASE_USERNAME,
                process.env.DATABASE_PASSWORD,
                {
                    dialect: "mysql",
                    host: process.env.DATABASE_HOST,
                    port: process.env.DATABASE_PORT,
                    logging: true,
                    define: {
                        freezeTableName: true
                    }
                }
            );
            await sequelize.authenticate();
            console.log("DB Authenticated Successfully");

            this.db.sequelize = sequelize;
            this.db.Sequelize = Sequelize;
            this.db.Op = Op;

            this.initDBTables(this.db);

            if (process.env.DATABASE_DDL_GENERATE == "CREATE") await sequelize.sync({ force: true });
            else if (process.env.DATABASE_DDL_GENERATE == "UPDATE") await sequelize.sync({ alter: true });

            this.isConnectionCreated = true;
        } catch (err) {
            this.isConnectionCreated = false;
            this.errorMessage = err;
            console.log("Error while Authentication ", err);
        }
    },
    initDBTables: function (db) {
        return true;
    }
}
