// serverless-base-util/response/Response.js
module.exports.Response = function () {
    this.isSendResponseAsFile = false;
    this.fileName = null;
    this.filePath = null;

    this.statusCode = null;
    this.headers = null;
    this.body = null;
    this.pagination = null;

    this.getBody = function () {
        if (this.pagination) {
            return { _metadata: this.pagination, records: this.body };
        } else {
            return this.body;
        }
    }

    this.setStatus = function (statusCode) {
        this.statusCode = statusCode;
        return this;
    }
    this.setHeader = function (key, value) {
        if (!this.headers) this.headers = {};
        this.headers[key] = value;
        return this;
    }
    this.setHeaders = function (headers) {
        this.headers = headers;
        return this;
    }
    this.setBody = function (body) {
        this.body = body;
        return this;
    }
    this.setPagination = function (pagination) {
        this.pagination = pagination;
        return this;
    }

    this.sendFile = function (fileName, filePath) {
        this.isSendResponseAsFile = true;
        this.fileName = fileName;
        this.filePath = filePath;
        return this;
    }

    return this;
};
